### PR TITLE
Adds MOSART-heat

### PR DESCRIPTION
**Pull request by liho745**
Added a new submodule to MOSART, MOSART-heat, which simulates water 
temperature in the rivers. MOSART-heat is coupled to both ELM and EAM. 
More specifically, MOSART-heat needs the runoff and soil temperature 
simulated by ELM, and solar radiation, wind speed etc. from EAM. 
To couple MOSART-heat with EAM, we newly added the ATM2ROF 
mapping (including code changes and several new ATM2ROF mapping files) 
which did not exist previously in E3SM. MOSART-heat can be turned on/off
with a heatflag option in user_nl_mosart.  

[BFB]
[NML]

### CODE REVIEW

**Comment by liho745** [comment]
I've reverted the config_machines change. 
I edited the top-level description with more details, including why adding ATM2ROF mapping. 
I did not, however, change rrtmpg submodule though. Perhaps I did not update all the sub-modules before committing my changes? In any case, I agree it should not be changed.

**Comment by thurber** [comment]
i pushed a new commit that reverts the `rrtmpg` submodule back it's status in `master`

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+! !PUBLIC MEMBER FUNCTIONS:
+    contains
+    
+    subroutine hillslopeHeat(iunit, theDeltaT)
+    ! !DESCRIPTION: calculate the temperature of surface and subsurface runoff.
+        implicit none
+        
+        integer, intent(in) :: iunit
+        real(r8), intent(in) :: theDeltaT        
+        !if(TUnit%fdir(iunit) >= 0 .and. TUnit%areaTotal(iunit) > TINYVALUE1) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+            THeat%Tqsur(iunit) = THeat%Tqsur(iunit)
+            ! adjust surface runoff temperature estimated based on the top-layer soil temperature, i.e., no less than freezing point
+            if(THeat%Tqsur(iunit) < 273.15_r8-TINYVALUE1) then
+                !write(unit=1821,fmt=*)  'Tqsub error at hillslopeHeat, ', iunit, THeat%Tqsur(iunit), THeat%Tqsub(iunit)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+                THeat%Tqsur(iunit) = 273.15_r8
+            !else
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+                !write(unit=1822,fmt=*)  'Tqsur good at hillslopeHeat, ', iunit, THeat%Tqsur(iunit), THeat%Tqsub(iunit)
+            end if
+            
+            THeat%Tqsub(iunit) = THeat%Tqsub(iunit)
+            if(THeat%Tqsub(iunit) < 273.15_r8-TINYVALUE1) then
+                !write(unit=1822,fmt=*)  'Tqsub error at hillslopeHeat, ', iunit, THeat%Tqsur(iunit), THeat%Tqsub(iunit)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+                THeat%Tqsub(iunit) = 273.15_r8
+            end if
+        !end if
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+    
+    subroutine subnetworkHeat(iunit, theDeltaT)
+    ! !DESCRIPTION: calculate the net heat balance of subnetwork channel.
+    use shr_sys_mod , only : shr_sys_flush
+        implicit none
+        integer, intent(in) :: iunit
+        real(r8), intent(in) :: theDeltaT    
+        
+        real(r8) :: Qsur, Qsub ! flow rate of surface and subsurface runoff separately
+        !if(TUnit%fdir(iunit) >= 0 .and. TUnit%areaTotal(iunit) > TINYVALUE1 .and. TUnit%tlen(iunit) >= TINYVALUE1) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+    subroutine mainchannelHeat(iunit, theDeltaT)
+    ! !DESCRIPTION: calculate the net heat balance of main channel.
+        use shr_sys_mod , only : shr_sys_flush
+        implicit none
+        integer, intent(in) :: iunit
+        real(r8), intent(in) :: theDeltaT    
+        integer    :: k
+        real(r8) :: Ha_temp, Ha_temp1, Ha_temp2
+        THeat%Ha_rin(iunit) = 0._r8
+        !if(TUnit%fdir(iunit) >= 0 .and. TUnit%areaTotal(iunit) > TINYVALUE1 .and. TUnit%rlen(iunit) >= TINYVALUE1) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+                else
+                    TRunoff%rarea(iunit,nt_nliq) = WaterAreaRatio*TUnit%rwidth(iunit) * TUnit%rlen(iunit)
+                end if
+                THeat%Hs_r(iunit) = cr_swrad(THeat%forc_solar(iunit), TRunoff%rarea(iunit,nt_nliq))
+                THeat%Hl_r(iunit) = cr_lwrad(THeat%forc_lwrad(iunit), THeat%Tr(iunit), TRunoff%rarea(iunit,nt_nliq))
+                THeat%He_r(iunit) = cr_latentheat(THeat%forc_t(iunit), THeat%forc_pbot(iunit), THeat%forc_vp(iunit), THeat%forc_wind(iunit), 1._r8, THeat%Tr(iunit), TRunoff%rarea(iunit,nt_nliq))
+                THeat%Hh_r(iunit) = cr_sensibleheat(THeat%forc_t(iunit), THeat%forc_pbot(iunit), THeat%forc_wind(iunit), 1._r8, THeat%Tr(iunit), TRunoff%rarea(iunit,nt_nliq))
+                THeat%Hc_r(iunit) = cr_condheat(THeat%Hs_r(iunit),THeat%Hl_r(iunit),TRunoff%rarea(iunit,nt_nliq))
+                
+                !do k=1,TUnit%nUp(iunit)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+                !    THeat%Ha_rin(iunit) = THeat%Ha_rin(iunit) - THeat%Ha_rout(TUnit%iUp(iunit,k))
+                !end do
+                
+                THeat%Ha_rin(iunit) = THeat%Ha_rin(iunit) - THeat%Ha_eroutUp(iunit) 
+                !if(TUnit%indexDown(iunit) > 0) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+                    THeat%Ha_rout(iunit) = -cr_advectheat(abs(TRunoff%erout(iunit,nt_nliq)+TRunoff%erout(iunit,nt_nice)), THeat%Tr(iunit))
+                !else
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+                !end if
+ 
+            ! change of energy due to heat exchange with the environment
+            THeat%deltaH_r(iunit) = theDeltaT * (THeat%Hs_r(iunit) + THeat%Hl_r(iunit) + THeat%He_r(iunit) + THeat%Hc_r(iunit) + THeat%Hh_r(iunit))
+            ! change of energy due to advective heat fluxes. Note here the advective heat flux is calculated differently from that by Huan Wu or van Vliet et al.
+            ! Their routing model is based on source-to-sink, while our model is explicitly tracing inflow from each upstream channel.
+            Ha_temp = cr_advectheat(TRunoff%erin(iunit,nt_nliq)+TRunoff%erin(iunit,nt_nice)+TRunoff%erlateral(iunit,nt_nliq)+TRunoff%erlateral(iunit,nt_nice),THeat%Tr(iunit))
+            THeat%deltaM_r(iunit) = theDeltaT * (THeat%Ha_lateral(iunit) + THeat%Ha_rin(iunit) - Ha_temp)
+            
+            !if(iunit == 85088) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+        real(r8) :: Ttmp1, Ttmp2  !
+        
+            if((TRunoff%wr(iunit,nt_nliq)+TRunoff%wr(iunit,nt_nice)) > TINYVALUE1 .and. THeat%forc_t(iunit) > 200._r8) then
+                Mr = TRunoff%wr(iunit,nt_nliq) * denh2o + TRunoff%wr(iunit,nt_nice) * denice
+                THeat%Tr(iunit) = THeat%Tr(iunit) + (THeat%deltaH_r(iunit)+THeat%deltaM_r(iunit)) / (Mr * cpliq)
+            else
+                THeat%Tr(iunit) = THeat%Tt(iunit)
+            end if
+            
+            !if(THeat%Tr(iunit) < 200._r8) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 13]
```diff
@@ -0,0 +1,570 @@
...
+    end subroutine mainchannelTemp_simple
+
+    subroutine reservoirHeat(iunit, theDeltaT)
+    ! !DESCRIPTION: calculate the net heat balance of reservoir. invoked after the regulation subroutine
+    ! simplified version as of 09/2014, now assuming the flow regulation won't modify the release water temperature directly
+    ! to be extended later, e.g., incorporating the stratification processes
+        implicit none
+        integer, intent(in) :: iunit
+        real(r8), intent(in) :: theDeltaT
+        !if(TUnit%indexDown(iunit) > 0) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out line

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 14]
```diff
@@ -0,0 +1,570 @@
...
+            THeat%Ha_rout(iunit) = -cr_advectheat(abs(TRunoff%erout(iunit,nt_nliq)+TRunoff%erout(iunit,nt_nice)), THeat%Tr(iunit))
+            !THeat%Ha_rout(iunit) = THeat%Ha_rout(iunit) - cr_advectheat(abs(TRunoff%erout(iunit,nt_nliq)+TRunoff%erout(iunit,nt_nice)), THeat%Tr(iunit))
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_heat_mod.F90 [hunk 15]
```diff
@@ -0,0 +1,570 @@
...
+      integer :: IDlist(1:4) = (/1233,1244,1138,1583/)
+      integer :: ios,ii                    ! flag of io status
+            
+
+      write(unit=nio,fmt="(12(e20.11))") rtmCTL%runofflnd_nt1(IDlist(1)), rtmCTL%templand_Tchanr_nt1(IDlist(1)), rtmCTL%templand_Ttrib_nt1(IDlist(1)), &
+                                         rtmCTL%runofflnd_nt1(IDlist(2)), rtmCTL%templand_Tchanr_nt1(IDlist(2)), rtmCTL%templand_Ttrib_nt1(IDlist(2)), &
+                                         rtmCTL%runofflnd_nt1(IDlist(3)), rtmCTL%templand_Tchanr_nt1(IDlist(3)), rtmCTL%templand_Ttrib_nt1(IDlist(3)), &
+                                         rtmCTL%runofflnd_nt1(IDlist(4)), rtmCTL%templand_Tchanr_nt1(IDlist(4)), rtmCTL%templand_Ttrib_nt1(IDlist(4))
+      
+      !write(unit=nio,fmt="(16(e20.11))") TRunoff%erout(IDlist(1),1), TRunoff%wr(IDlist(1),1), THeat%Tt(IDlist(1)), THeat%Tr(IDlist(1)), &
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_heat_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 16]
```diff
@@ -2387,7 +2408,17 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
+       if(heatflag) then
+         do n = rtmCTL%begr,rtmCTL%endr
+             if (rtmCTL%mask(n) .eq. 1 .or. rtmCTL%mask(n) .eq. 3) then
+                 rtmCTL%templand_Tqsur(n) = rtmCTL%templand_Tqsur(n) + THeat%Tqsur(n)
+                 rtmCTL%templand_Tqsub(n) = rtmCTL%templand_Tqsub(n) + THeat%Tqsub(n)
+                 rtmCTL%templand_Ttrib(n) = rtmCTL%templand_Ttrib(n) + THeat%Tt_avg(n)
+                 rtmCTL%templand_Tchanr(n) = rtmCTL%templand_Tchanr(n) + THeat%Tr_avg(n)
+             end if
+         enddo
+	   end if
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please indent the code

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 17]
```diff
@@ -2431,6 +2462,28 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
+            rtmCTL%templand_Ttrib(n) = rtmCTL%templand_Ttrib(n) / float(nsub)
+            rtmCTL%templand_Tchanr(n) = rtmCTL%templand_Tchanr(n) / float(nsub)
+         else
+            rtmCTL%templand_Tqsur(n) = spval
+            rtmCTL%templand_Tqsub(n) = spval
+            rtmCTL%templand_Ttrib(n) = spval
+            rtmCTL%templand_Tchanr(n) = spval
+         end if
+      end do
+	end if
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please indent the code

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 18]
```diff
@@ -3856,7 +3909,106 @@ subroutine MOSART_init
         endif
      end if
 !#endif
-
+     
+	 if(heatflag) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please indent the code

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 19]
```diff
@@ -3856,7 +3909,106 @@ subroutine MOSART_init
...
+        allocate (THeat%Hc_t(begr:endr))
+        THeat%Hc_t = 0._r8
+        allocate (THeat%deltaH_t(begr:endr))
+        THeat%deltaH_t = 0._r8
+        allocate (THeat%deltaM_t(begr:endr))
+        THeat%deltaM_t = 0._r8
+
+        allocate (THeat%Tr(begr:endr))
+        THeat%Tr = 273.15_r8
+        !allocate (THeat%Ha_r(begr:endr))
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please delete commented out line

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 20]
```diff
@@ -3856,7 +3909,106 @@ subroutine MOSART_init
...
+        THeat%deltaM_r = 0._r8
+
+        allocate (THeat%Tt_avg(begr:endr))
+        THeat%Tt_avg = 273.15_r8
+        allocate (THeat%Tr_avg(begr:endr))
+        THeat%Tr_avg = 273.15_r8
+		
+       ! read the parameters for mosart-heat
+        if(endr >= begr) then
+            !call check_ret(nf_open(frivinp_rtm, 0, ncid), 'Reading file: ' // frivinp_rtm)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please delete commented out lines

File: cime/src/drivers/mct/main/prep_rof_mod.F90 [hunk 21]
```diff
@@ -209,7 +280,40 @@ subroutine prep_rof_accum(timer)
...
-  end subroutine prep_rof_accum
+  end subroutine prep_rof_accum_lnd
+
+  !================================================================================================
+
+  subroutine prep_rof_accum_atm(timer)
+
+    !---------------------------------------------------------------
+    ! Description
+    ! Accumulate land input to river component
```
Comment (bishtgautam) [inline cime/src/drivers/mct/main/prep_rof_mod.F90]: Is this comment correct? Shouldn't it be atm instead of land?

**Comment by thurber** [comment]
@liho745, @bishtgautam, any remaining next steps to get this merged?

**Comment by bishtgautam** [comment]
@thurber, The MPAS submodule has been incorrectly updated in this PR and there are a bunch of conflicts. Once #3074 is merged to master, this PR will be rebased off the latest master.

**Comment by jonbob** [comment]
@bishtgautam - With help from @jedwards4b I think we've fixed the issue about missing valid mapping files, but it will need some testing. With a test merge, this passed:
* SMS.T62_oQU240.CMPASO-NYF.anvil_intel
* SMS.ne4_oQU240.A_WCYCL1850.anvil_intel
so one test with MOSART and one without. But this PR probably needs to be rebased -- the merge seemed complicated and I had changes to one file:

```cime/src/drivers/mct/cime_config/config_component.xml```

that was very different between master and this branch...

**Comment by bishtgautam** [comment]
Thanks, @jonbob

**Comment by rljacob** [comment]
@golaz do you want this for v2 watercycle?

**Comment by golaz** [comment]
If it's BFB, it could be merged in the code. But turning it on in v2 will depend on someone volunteering to make a simulation with it and explaining the benefits of the feature (and demonstrating no degradation).

**Comment by dqwu** [comment]
@bishtgautam 
This PR has caused 21 tests failed with latest next branch:
https://my.cdash.org/viewTest.php?onlyfailed&buildid=1719105

I have manually run ERS.f19_f19.ICLM45 on anlworkstation, and the stack trace is:
```
[0] Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
[0] 
[0] Backtrace for this error:
[0] #0  0x7f2c5b2becaf in ???
[0] #1  0x11e7da9 in __mosart_physics_mod_MOD_euler
[0] 	at components/mosart/src/riverroute/MOSART_physics_mod.F90:133
[0] #2  0x1174c64 in __rtmmod_MOD_rtmrun
[0] 	at components/mosart/src/riverroute/RtmMod.F90:2246
[0] #3  0x114c50b in __rof_comp_mct_MOD_rof_run_mct
[0] 	at components/mosart/src/cpl/rof_comp_mct.F90:361
[0] #4  0x427451 in __component_mod_MOD_component_run
[0] 	at cime/src/drivers/mct/main/component_mod.F90:714
[0] #5  0x4115b4 in __cime_comp_mod_MOD_cime_run
[0] 	at cime/src/drivers/mct/main/cime_comp_mod.F90:2640
[0] #6  0x42611f in cime_driver
[0] 	at cime/src/drivers/mct/main/cime_driver.F90:133
[0] #7  0x42623d in main
[0] 	at cime/src/drivers/mct/main/cime_driver.F90:23
```

**Comment by bishtgautam** [comment]
I'm reverting this PR again and will ask the developers to dig into these errors.

**Comment by jonbob** [comment]
With the recent commit, a test merge now passes:
* SMS_D.ne4_oQU240.A_WCYCL1850.anvil_gnu
* ERS.f19_f19.ICLM45.anvil_gnu

It may make sense to rebase to master since the merge has conflicts. I can work on that next.

**Comment by jonbob** [comment]
@bishtgautam - it may be easier to simply merge this PR instead of rebasing it. Do you want me to take it on? I think it does now pass some of the tests that had been failing

**Comment by bishtgautam** [comment]
I agree with you regarding rebasing this PR. I will (re-re-)merge this PR to next later today and double check my work before pushing it to next.

**Comment by jonbob** [comment]
@bishtgautam - sounds good, and thanks

**Comment by bishtgautam** [comment]
Thanks to @jonbob for fixing the bug found by gnu, this PR has been merged to master.

**Comment by hydrotian** [comment]
Bravo! Thanks to @jonbob and @bishtgautam !

**Comment by liho745** [comment]
Big Thanks to all of you!

**Pull request merged**